File: ts-mcp-client/src/debug-bus.ts
/**
 * Every JSON-RPC frame the MCP client sends/receives, plus lifecycle and bridge
 * events, flow through this bus. The /debug/stream SSE endpoint relays them to the
 * frontend so each capability page can show what is happening on the wire.
 */
export type FrameKind =
  | 'request'
  | 'response'
  | 'error'
  | 'notification'
  | 'lifecycle'
  | 'elicitation'
  | 'note';

export interface Frame {
  seq: number;
  ts: number;
  dir: 'send' | 'recv' | 'local';
  kind: FrameKind;
  method?: string;
  id?: string | number | null;
  summary?: string;
  payload?: unknown;
  /** Correlates frames produced while handling one frontend-triggered action. */
  trace?: string;
}

type FrameListener = (frame: Frame) => void;

/**
 * A tiny edge-safe event bus (no `node:events`). Frames fan out to every
 * registered listener; the `/debug/stream` SSE endpoint subscribes via on/off.
 */
class DebugBus {
  private seq = 0;
  private readonly listeners = new Set<FrameListener>();

  on(_event: 'frame', listener: FrameListener): this {
    this.listeners.add(listener);
    return this;
  }
  off(_event: 'frame', listener: FrameListener): this {
    this.listeners.delete(listener);
    return this;
  }
  emitFrame(f: Omit<Frame, 'seq' | 'ts'>): Frame {
    const frame: Frame = { ...f, seq: ++this.seq, ts: Date.now() };
    for (const listener of [...this.listeners]) {
      try {
        listener(frame);
      } catch {
        // a listener must not break the bus
      }
    }
    return frame;
  }
}

export const bus = new DebugBus();

File: py-mcp-client/debug_bus.py
"""The wire-debug frame bus.

Every JSON-RPC frame the MCP client sends/receives, plus lifecycle and bridge events,
flow through this bus. The ``/debug/stream`` SSE endpoint relays them to the frontend so
each capability page can show what is happening on the wire. Thread-safe: frames are
emitted from the SDK transport's request/subscription threads while the SSE relay runs
on the event loop.
"""

from __future__ import annotations

import threading
import time
from collections.abc import Callable

#: The Frame shape the frontend expects: seq, ts, dir, kind, method?, id?, summary?, payload?, trace?.
FrameListener = Callable[[dict], None]


class DebugBus:
  """A tiny thread-safe event bus; the ``/debug/stream`` endpoint subscribes via on/off."""

  def __init__(self) -> None:
    self._seq = 0
    self._listeners: set[FrameListener] = set()
    self._lock = threading.Lock()

  def on(self, listener: FrameListener) -> None:
    with self._lock:
      self._listeners.add(listener)

  def off(self, listener: FrameListener) -> None:
    with self._lock:
      self._listeners.discard(listener)

  def emit_frame(self, partial: dict) -> dict:
    """Stamp ``seq`` + ``ts`` and fan the frame out to every registered listener."""
    with self._lock:
      self._seq += 1
      frame = {**partial, "seq": self._seq, "ts": int(time.time() * 1000)}
      listeners = list(self._listeners)
    for listener in listeners:
      try:
        listener(frame)
      except Exception:  # noqa: BLE001 — a listener must not break the bus
        pass
    return frame


bus = DebugBus()
